### PR TITLE
Generate a manifest if one is not provided.

### PR DESCRIPTION
**Pull request by dlorenc**
Note: this also includes the change from #119. I pulled that out into it's own PR because this one will likely take longer to get merged.

I mainly did this as a prototype, but I actually like it now. LMK what you think.

### CODE REVIEW

File: examples/no-manifest/skaffold.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  # artifacts is a list of the actual images you're going to be building
+  # you can include as many as you want here.
+  artifacts:
+    # The name of the image we're going to build, the full reference to the built
+    # image will be available later in the the deploy section
+  - imageName: changeme # gcr.io/k8s-skaffold/skaffold-example
+    # dockerfilePath: Dockerfile # not required if its just $workspace/Dockerfile
+
+    # workspace is the path to your dockerfile context
+    workspace: examples/getting-started
```
Comment (r2d4) [inline examples/no-manifest/skaffold.yaml]: examples/no-manifest

File: examples/no-manifest/skaffold.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  # gitCommit - tags the image with the git commit of your current repository
+  
+  # tagPolicy: sha256
+
+  # artifacts is a list of the actual images you're going to be building
+  # you can include as many as you want here.
+  artifacts:
+    # The name of the image we're going to build, the full reference to the built
+    # image will be available later in the the deploy section
+  - imageName: changeme # gcr.io/k8s-skaffold/skaffold-example
```
Comment (r2d4) [inline examples/no-manifest/skaffold.yaml]: you can just put this as `gcr.io/k8s-skaffold/skaffold-example`

File: examples/no-manifest/skaffold.yaml [hunk 3]
```diff
@@ -0,0 +1,47 @@
+apiVersion: skaffold/v1
+kind: Config
+build:
+  # tagPolicy determines how skaffold is going to tag your image
```
Comment (r2d4) [inline examples/no-manifest/skaffold.yaml]: you can remove all the comments

File: pkg/skaffold/deploy/kubectl.go [hunk 4]
```diff
@@ -31,6 +36,35 @@ import (
 	"github.com/pkg/errors"
 )
 
+// Slightly modified from kubectl run --dry-run
+var deploymentTemplate = template.Must(template.New("deployment").Parse(`apiVersion: extensions/v1beta1
+kind: Deployment
```
Comment (r2d4) [inline pkg/skaffold/deploy/kubectl.go]: you need a service too right?

File: pkg/skaffold/deploy/kubectl.go [hunk 5]
```diff
@@ -44,6 +78,23 @@ func NewKubectlDeployer(cfg *config.DeployConfig) (*KubectlDeployer, error) {
 // Run templates the provided manifests with a simple `find and replace` and
 // runs `kubectl apply` on those manifests
 func (k *KubectlDeployer) Run(out io.Writer, b *build.BuildResult) (*Result, error) {
+
```
Comment (r2d4) [inline pkg/skaffold/deploy/kubectl.go]: nit remove newline

File: examples/no-manifest/skaffold.yaml (hunk 1 above)
Comment (r2d4) [reply examples/no-manifest/skaffold.yaml]: on second thought, you can leave this like it is and remove the dockerfile and main.go if you want. Then this folder can just be the skaffold.yaml

File: pkg/skaffold/docker/parse.go [hunk 6]
```diff
@@ -108,6 +109,47 @@ func GetDockerfileDependencies(workspace string, r io.Reader) ([]string, error)
...
+		return nil, errors.Wrap(err, "parsing dockerfile")
+	}
+
+	// Check the dockerfile and the base.
+	ports := []string{}
+	for _, value := range res.AST.Children {
+		switch value.Value {
+		case from:
+			base := value.Next.Value
+			if base == "SCRATCH" {
```
Comment (r2d4) [inline pkg/skaffold/docker/parse.go]: why couldn't you expose ports in a scratch image?

Comment (r2d4) [reply pkg/skaffold/docker/parse.go]: you don't have ca-certs but internal services, etc.

File: pkg/skaffold/deploy/kubectl.go (hunk 4 above)
Comment (dlorenc) [reply pkg/skaffold/deploy/kubectl.go]: You don't *need* one, only if you want to access it over an LB or something. I think we should probably recommend port-forward in any samples.

File: pkg/skaffold/docker/parse.go (hunk 6 above)
Comment (dlorenc) [reply pkg/skaffold/docker/parse.go]: It's about the FROM check - we check your Dockerfile and the base image for ports. a SCRATCH base image won't have any PORTS exposed itself so we don't need to try to check it. We still check your Dockerfile for EXPOSES though.

**Comment by r2d4** [comment]
TIL switching the repository from private to public breaks the existing forks (why? I have no clue).  I had to re-fork and push to the new remote

https://help.github.com/articles/what-happens-to-forks-when-a-repository-is-deleted-or-changes-visibility/#changing-a-private-repository-to-a-public-repository

**Comment by dlorenc** [comment]
> TIL switching the repository from private to public breaks the existing forks (why? I have no clue). I had to re-fork and push to the new remote

Yeah I'll have to recreate that, but I think this one is still working, unless I'm missing something.

File: pkg/skaffold/deploy/kubectl.go (hunk 5 above)
Comment (dlorenc) [reply pkg/skaffold/deploy/kubectl.go]: done.

**Comment by dlorenc** [comment]
> Yeah I'll have to recreate that, but I think this one is still working, unless I'm missing something.

nvm, it wasn't picking up new commits. I see. Opened a new PR: https://github.com/GoogleCloudPlatform/skaffold/pull/141

**Pull request closed without merging**